### PR TITLE
[InstCombine] Add ctpop(A | B) + ctpop(A & B) -> ctpop(A) + ctpop(B)

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1733]:
```suggestion
  if (match(&I, m_c_Add(m_OneUse(m_Intrinsic<Intrinsic::ctpop>(
                       m_And(m_Value(A), m_Value(B))))), mOneUse(m_Intrinsic<Intrinsic::ctpop>(
                       m_c_Or(m_Specific(A), m_Specific(B)))))))) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1721]: Alive2 proofs?

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1737]:
```suggestion
    return BinaryOperator::CreateAdd(
               Builder.CreateUnaryIntrinsic(Intrinsic::ctpop, A),
               Builder.CreateUnaryIntrinsic(Intrinsic::ctpop, B));
```

File: llvm/test/Transforms/InstCombine/ctpop.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/ctpop.ll:267]: Please add some negative tests.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1721]: https://alive2.llvm.org/ce/z/NmOVOO

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp]: Tried this and now the transforms don't work anymore

**Comment by AtariDreams** [comment]
@dtcxzyw Done!

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp]: > Tried this and now the transforms don't work anymore

Oh, you need `m_Deferred` instead of `m_Specific`. Sorry for my mistake.

**Comment by DavidSpickett** [comment]
(since you asked about this on Discord)

If you click through to the results and click a whole bunch more (it's not super easy to find), you'll get to the end of the log and see:
```
MT: command "C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe /nologo /manifest unittests\TargetParser\TargetParserTests.exe.manifest /outputresource:unittests\TargetParser\TargetParserTests.exe;#1" failed (exit code 0x1f) with the following output:
mt.exe : general error c101008d: Failed to write the updated manifest to the resource of file "unittests\TargetParser\TargetParserTests.exe". Operation did not complete successfully because the file contains a virus or potentially unwanted software.
```

Which is not related to your patch. Ideally you'd get a clean run at some point, maybe by rebasing once the patch is approved, but if that's not possible maybe your reviewers will allow a merge without a Windows CI run.

**Comment by dtcxzyw** [comment]
@AtariDreams Does this pattern exist in some real-world applications?

**Comment by AtariDreams** [comment]
> @AtariDreams Does this pattern exist in some real-world applications?

Given the popcount(A|B) does exist, it makes sense that popcount(A&B) + popcount(A|B) exists.

**Comment by dtcxzyw** [comment]
The implementation LGTM. But I am sorry I cannot give the approval unless you can prove that it benefits some real-world applications.

There are an infinite number of atomic simplifications. But obviously you cannot handle all of them in LLVM.

**Comment by nikic** [comment]
I agree with @dtcxzyw. The pattern here is simple enough that we *could* accept it without knowing about a specific use-case, but at the same time I have a really hard time imaging where this could possibly come up. I'm going to decline it for now.

**Pull request closed without merging**
